File: framework/languages/java/e2e/ResilienceLifecycle/Client/src/main/java/systems/zlink/e2e/resiliencelifecycle/client/Scenarios/RlC2TopologyRecoveryScenario.java
package systems.zlink.e2e.resiliencelifecycle.client.Scenarios;

import systems.zlink.e2e.resiliencelifecycle.client.Support.ResilienceScenarioContext;

public final class RlC2TopologyRecoveryScenario {
    private RlC2TopologyRecoveryScenario() {
    }

    public static void run(ResilienceScenarioContext context) {
        context.waitForTopology(2);
        context.signal("c2-ready");
        context.waitForSignal("c2-crashed");
        context.waitForTopologyWithout("api-b", 30);
        context.collectStableProvidersWithoutFailures("c2-after-crash", "api-b", "api-a");
        context.signal("c2-survivor-observed");
        context.waitForSignal("c2-restored");
        context.waitForTopologyEndpoint("api-b", context.options().apiBEndpoint());
        context.driveUntilEvidence(
            context.adminB(), "c2-restored", "RL-C2 restored provider traffic missing");
        System.out.println("scenario RL-C2 passed");
    }
}

File: framework/languages/java/e2e-kotlin/ResilienceLifecycle/Client/src/main/kotlin/systems/zlink/e2e/kotlin/resiliencelifecycle/Scenarios/RlC2TopologyRecoveryScenario.kt
package systems.zlink.e2e.kotlin.resiliencelifecycle

import java.time.Duration

fun ClientScenarioContext.runTopologyRecoveryScenario() {
    waitForTopology(2)
    val before = requestWork("rl-c2-before-crash", Duration.ofSeconds(5))
    ensure(before.value() == "work:rl-c2-before-crash", "RL-C2 before-crash payload mismatch")
    waitForEvidenceValueAny("WorkReq", "rl-c2-before-crash", adminA(), adminB())

    signal("c2-ready")
    waitForSignal("c2-crashed")
    waitForTopologyMissing("api-b")

    repeat(8) { index ->
        val reply = requestWork("rl-c2-after-crash-$index", Duration.ofSeconds(5))
        ensure(reply.providerRid() == "api-a", "RL-C2 request used stale provider ${reply.providerRid()}")
        ensure(reply.value() == "work:rl-c2-after-crash-$index", "RL-C2 after-crash payload mismatch")
    }
    waitForEvidenceValueAny("WorkReq", "rl-c2-after-crash-0", adminA())

    signal("c2-restart")
    waitForSignal("c2-restarted")
    waitForTopology(2)
    waitUntilProviderServesAfterRecovery("api-b", "rl-c2-restored")
    println("scenario RL-C2 passed")
}

private fun ClientScenarioContext.waitUntilProviderServesAfterRecovery(providerRid: String, prefix: String) {
    repeat(40) { attempt ->
        val value = "$prefix-$attempt"
        try {
            val reply = requestWork(value, Duration.ofSeconds(5))
            ensure(reply.value() == "work:$value", "RL-C2 restored provider payload mismatch")
            if (reply.providerRid() == providerRid) {
                waitForEvidenceValueAny("WorkReq", value, adminB())
                return
            }
        } catch (_: RuntimeException) {
        }
        sleep(250)
    }
    throw IllegalStateException("RL-C2 restored provider $providerRid did not receive traffic")
}
